kythe/docs/kythe-storage.txt

```diff
@@ -240,12 +241,24 @@ The initial definition of a VName will include the following 5 fields:
    path can't be represented without "escaping" from its designated root
    directory, it's usually a good sign that a separate Root label should be
    assigned.
+   +
+   +
+   The components of a path are separated by "/"s, even if the
+   underlying operating system uses a different separator. The path
+   must not start with a "/" -- that is, it is relative to the *Root*.
+   If the root is empty, it is treated as meaning the top of the
+   corpus (e.g., the same as "/" if the corpus maps directly to the
+   file system).
```

Comment:
VName paths are not necessarily file system paths.  They should generally *map* to them, but that mapping is not guaranteed to be direct.  This is independent of root.  We do generally try to maintain the semantics that an empty root corresponds to a concrete file in the corpus (as opposed to a generated or other build artifact), but it is not guaranteed that `(cd ${VNAME[corpus]}; cat ${VNAME[path]})` will do anything meaningful.

Reply:
I was trying to clarify what "path-structured label" means -- namely a list of names separated by "/"s. 

Is that correct? If so, I'll generalize the "mapping" language to avoid implying that there must be a file system.

(I don't think we need to introduce any concept of hierarchy, such as the Unix or Windows file systems; but it might be desirable in future because sometimes it seems to make sense to point to a "directory" rather than a file, e.g. for Python's `from foo import *`. But that should probably be left to a separate discussion because it has wider implications.)

Reply:
It looks like the closest thing we have to a definition right now is in the [Kythe URI spec](https://kythe.io/docs/kythe-uri-spec.html). I agree it'd make sense for them to be in sync, even if only by reference.

Reply:
The URI seems to allow extra "/"s in the path and doesn't say anything about "." and "..". Is this intentional or an oversight? My guess is that it's intentional because we should allow non-canonical queries; but the vnames should be in a canonical form to make "joins" work with simple string matching.

I'll add a pointer to the URI spec (and pct-encoding isn't needed for paths, assuming that they're UTF-8) and also add some language about "cleaning", "." and "..".

Reply:
Yeah, agreed, the URI spec has some URI-specific bits (like the escaping) that don't really apply to the abstract VName. 

The more I think about it, the more I think maybe these kinds of rules belong in the schema rather than the data format docs. As you pointed out, the format (purposely) allows non-canonical data, but in the schema we may implicitly rely on more constraints.

Reply:
I don't want to get involved with the structure of Kythe's documentation, so I've made what I hope are minimalist changes.

I think we need to clarify where normalization is done -- e.g., when Kythe facts are output by an indexer, can they be denormalized (containing ".", "..", in the Path, for example) or must the indexer normalize them? Similarly, can queries be denormalized?

BTW, Python has [posix.path.normpath](https://docs.python.org/3/library/os.path.html#os.path.normpath); presumably similar functions exist in C++, Java, Go, etc.